### ISSUE TITLE
Aggregate pushdown

### ISSUE THREAD

**Issue by jackvincenow** [hyrise/hyrise#1943]
Hi team,

Any plans on implementing aggregation pushdowns into the LPQ optimizer?

for example, we have a query that applies groupby + agg on a sub query which performs a join.
The group by + agg depend on both tables, but the join on the original table is expensive.
Instead, we could rewrite the SQL to have the outer aggregations happening before the join, thus save a lot of time because the join is much quicker.

**Comment by mrks** [comment]
Do you have a sample original and rewritten SQL query or an LQP with drawings for us? I'm having a hard time coming up with an example.

**Comment by jackvincenow** [comment]
Here is a sample query, used in the ONTIME airlines public data set:

1. original query applies group/agg/having after the join:

SELECT Dest d, Name n, COUNT(*), AVG(delayMinutes) FROM ontime_full
JOIN airports_full on (airports_full.IATA = ontime_full.Dest)
GROUP BY b, n HAVING c > 500000 

2. optimized query replaces the "FROM ontime_full" part with a sub query that does the group by + agg + having before the join:

SELECT dest, Name n, c AS count_flights, avg_delay FROM 
(
	SELECT Dest dest, count(*) c, AVG(delayMinutes) avg_delay FROM ontime
	GROUP BY dest having c > 500000
) JOIN airports_full ON airports_full.IATA = dest

**Comment by Bouncner** [comment]
Do you know that there is a one-to-one relation between `airports_full.IATA` and `ontime_full.Dest` from the data set?

**Comment by mrks** [comment]
Got it. I reproduced your example and had a look at the LQP. Generally, this would be possible, but there are a couple of steps that need to happen:

1) We need to identify that `Name` is not an actual GROUP BY column as it is dependent on `Dest`. While #1929 is a first step into that direction, this example is more complex as we need to leverage foreign key information.
2) Once that has happened, we need a rule that goes over the AggregateNode's ANY() expressions and checks whether they have been added by a `n:1` join. In that case, we should be able to remove them from the AggregateNode.
3) The final part, pushing down the aggregate, is more or less trivial.

At least the first part is in planning, but as we do not have foreign key support yet, it will take a while. Honestly, I do not expect us to be able to do this optimization within the next months.

However, if you have an interest in getting this to work, we'll be happy to help.
